daemon/daemon/policy.go

```diff
@@ -195,91 +226,165 @@ func (d *Daemon) policyAdd(path string, node *policy.Node) (bool, error) {
 	return policyModified, nil
 }
 
-func (d *Daemon) PolicyAdd(path string, node *policy.Node) error {
+func (d *Daemon) PolicyAdd(path string, node *policy.Node) *apierror.ApiError {
 	log.Debugf("Policy Add Request: %s %+v", path, node)
 
 	if !strings.HasPrefix(path, common.GlobalLabelPrefix) {
-		return fmt.Errorf("the given path %q doesn't have the prefix %q", path, common.GlobalLabelPrefix)
+		return apierror.New(PutPolicyInvalidPathCode,
+			"Invalid path %s: must start with %s", path, common.GlobalLabelPrefix)
 	}
 
-	var (
-		err            error
-		policyModified bool
-	)
-
 	d.policy.Mutex.Lock()
-	defer func() {
-		d.policy.Mutex.Unlock()
-		if err == nil && policyModified {
-			log.Info("New policy received, triggering updates...")
-			d.triggerPolicyUpdates([]policy.NumericIdentity{})
+	defer d.policy.Mutex.Unlock()
+
+	if policyModified, err := d.policyAdd(path, node); err != nil {
+		return apierror.Error(PutPolicyFailureCode, err)
+	} else if policyModified {
+		if err = node.ResolveTree(); err != nil {
+			return apierror.Error(PutPolicyFailureCode, err)
 		}
-	}()
 
-	policyModified, err = d.policyAdd(path, node)
-	if err != nil || !policyModified {
-		return err
+		log.Info("New policy imported, regenerating...")
+		d.triggerPolicyUpdates([]policy.NumericIdentity{})
+	}
+
+	return nil
+}
+
+func (d *Daemon) deleteNode(node *policy.Node, parent *policy.Node) {
+	if node == d.policy.Root {
+		d.policy.Root = policy.NewNode(common.GlobalLabelPrefix, nil)
+		d.policy.Root.Path()
+	} else {
+		delete(parent.Children, node.Name)
 	}
-	err = node.ResolveTree()
-	return err
 }
 
 // PolicyDelete deletes the policy set in the given path from the policy tree. If
 // cover256Sum is set it finds the rule with the respective coverage that rule from the
 // node. If the path's node becomes ruleless it is removed from the tree.
-func (d *Daemon) PolicyDelete(path, cover256Sum string) (err error) {
+func (d *Daemon) PolicyDelete(path, cover256Sum string) *apierror.ApiError {
 	log.Debugf("Policy Delete Request: %s, cover256Sum %s", path, cover256Sum)
 
 	d.policy.Mutex.Lock()
-	defer func() {
-		d.policy.Mutex.Unlock()
-		if err == nil {
-			d.triggerPolicyUpdates([]policy.NumericIdentity{})
-		}
-	}()
+	defer d.policy.Mutex.Unlock()
```

Comment:
Do we need to have the policy tree locked if we perform `d.triggerPolicyUpdates`?